

# Part A

# And Part B


In [1]:



import numpy as np

import pandas as pd

heart_d_df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")

heart_d_df.head()


Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0   63      f       typical     145   233        2    150      2.3   
1   67      f  asymptomatic     160   286        2    108      1.5   
2   67      f  asymptomatic     120   229        2    129      2.6   
3   37      f    nonanginal     130   250        0    187      3.5   
4   41      m    nontypical     130   204        2    172      1.4   

         Thal  AHD  
0       fixed   No  
1      normal  Yes  
2  reversable  Yes  
3      normal   No  
4      normal   No


## Part C



In [2]:



hd_feature_cols = [ 'Age', 'RestBP', 'Chol', 'RestECG', 'MaxHR', 'Oldpeak']

HDFM = heart_d_df [hd_feature_cols]

LABEL_V = heart_d_df ['AHD']







# Part D



In [3]:


from sklearn.model_selection import train_test_split

HDFM_training, HDFM_testing, Label_training, Label_testing = train_test_split(HDFM, LABEL_V, test_size = 0.25, random_state = 6)







# Part E



In [4]:


from sklearn.linear_model import LogisticRegression

from sklearn.tree import DecisionTreeClassifier

from sklearn.neighbors import KNeighborsClassifier

from sklearn.metrics import accuracy_score

k_value = 3

rs_tree_value = 5

k3_knn_classifier = KNeighborsClassifier(n_neighbors=k_value)

rs5_decision_tree = DecisionTreeClassifier(random_state = rs_tree_value)

rs_log_regression = LogisticRegression()



k3_knn_classifier.fit(HDFM_training,Label_training)

rs5_decision_tree.fit(HDFM_training,Label_training)

rs_log_regression.fit(HDFM_training,Label_training)


label_predict_knn = k3_knn_classifier.predict(HDFM_testing)

label_predict_dec_tree = rs5_decision_tree.predict (HDFM_testing)

label_predict_log_reg = rs_log_regression.predict (HDFM_testing) 

acc_rate_knn_k3 = accuracy_score (Label_testing,label_predict_knn)

acc_rate_dec_tree = accuracy_score (Label_testing,label_predict_dec_tree)

acc_rate_log_red = accuracy_score (Label_testing,label_predict_log_reg)


print( "knn rate: " +  str(acc_rate_knn_k3))

print( "decision tree rate: " + str(acc_rate_dec_tree)) 

print ("logistic regression rate: " + str(acc_rate_log_red))






knn rate: 0.6447368421052632
decision tree rate: 0.618421052631579
logistic regression rate: 0.6710526315789473




# Part E Results

According to the accuracy scores: logistic regression had the best score, and decision tree had the worse score.






# Part F

Using pandas functions to do OneHotEncoding

In [5]:




heart_d_df_transformed = pd.get_dummies (heart_d_df, prefix_sep='', prefix ='', 
                                         columns= ['Gender', 'ChestPain', 'Thal'])

print (heart_d_df_transformed)




     Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  AHD  f  m  asymptomatic  \
0     63     145   233        2    150      2.3   No  1  0             0   
1     67     160   286        2    108      1.5  Yes  1  0             1   
2     67     120   229        2    129      2.6  Yes  1  0             1   
3     37     130   250        0    187      3.5   No  1  0             0   
4     41     130   204        2    172      1.4   No  0  1             0   
..   ...     ...   ...      ...    ...      ...  ... .. ..           ...   
296   45     110   264        0    132      1.2  Yes  1  0             0   
297   68     144   193        0    141      3.4  Yes  1  0             1   
298   57     130   131        0    115      1.2  Yes  1  0             1   
299   57     130   236        2    174      0.0  Yes  0  1             0   
300   38     138   175        0    173      0.0   No  1  0             0   

     nonanginal  nontypical  typical  fixed  normal  reversable  
0             0      


# Part G
# (doing part D and E with the transformed data in part F)



In [6]:


import numpy as np

import pandas as pd


heart_d_df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")


heart_d_df_transformed = pd.get_dummies (heart_d_df, prefix_sep='', prefix ='', 
                                         columns= ['Gender', 'ChestPain', 'Thal'])


heart_d_df_trans_cols = list(heart_d_df_transformed.columns)


heart_d_df_trans_cols.remove('AHD')


HDTFM = heart_d_df_transformed[heart_d_df_trans_cols]

LABEL_V = heart_d_df_transformed['AHD']



from sklearn.model_selection import train_test_split

HDTFM_training, HDTFM_testing, Label_training, Label_testing = train_test_split(HDTFM, LABEL_V, test_size = 0.25, random_state = 6)


from sklearn.linear_model import LogisticRegression

from sklearn.tree import DecisionTreeClassifier

from sklearn.neighbors import KNeighborsClassifier

from sklearn.metrics import accuracy_score


k_value = 3

rs_tree_value = 5


k3_knn_classifier = KNeighborsClassifier(n_neighbors=k_value)

rs5_decision_tree = DecisionTreeClassifier(random_state = rs_tree_value)

max_iter_value = 1067 # min value on this computer to not reach an error
rs_log_regression = LogisticRegression(max_iter=1500)


k3_knn_classifier.fit(HDTFM_training,Label_training)

rs5_decision_tree.fit(HDTFM_training,Label_training)

rs_log_regression.fit(HDTFM_training,Label_training)


label_predict_knn = k3_knn_classifier.predict(HDTFM_testing)

label_predict_dec_tree = rs5_decision_tree.predict(HDTFM_testing)

label_predict_log_reg = rs_log_regression.predict(HDTFM_testing) 


acc_rate_knn_k3 = accuracy_score (Label_testing,label_predict_knn)

acc_rate_dec_tree = accuracy_score (Label_testing,label_predict_dec_tree)

acc_rate_log_red = accuracy_score (Label_testing,label_predict_log_reg)



print( "knn rate: " +  str(acc_rate_knn_k3))

print( "decision tree rate: " + str(acc_rate_dec_tree)) 

print ("logistic regression rate: " + str(acc_rate_log_red))




knn rate: 0.6447368421052632
decision tree rate: 0.7368421052631579
logistic regression rate: 0.7763157894736842



# Part G Results

KNN with k = 3 did not change much in  accuracy results.

Decision Tree accuracy result improveed quite a bit.

Logistic Regression accuracy results improved quite a bit too.













# Part H



In [7]:


import numpy as np

import pandas as pd


heart_d_df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")

heart_d_df_transformed = pd.get_dummies (heart_d_df, prefix_sep='', prefix ='', 
                                         columns= ['Gender', 'ChestPain', 'Thal'])

heart_d_df_trans_cols = list(heart_d_df_transformed.columns)

heart_d_df_trans_cols.remove('AHD')


HDTFM = heart_d_df_transformed[heart_d_df_trans_cols]

LABEL_V = heart_d_df_transformed['AHD']



from sklearn.linear_model import LogisticRegression

from sklearn.tree import DecisionTreeClassifier

from sklearn.neighbors import KNeighborsClassifier



k_value = 3

rs_tree_value = 5


k3_knn_classifier = KNeighborsClassifier(n_neighbors=k_value)

rs5_decision_tree = DecisionTreeClassifier(random_state = rs_tree_value)

max_iter_value = 1067 # min value on this computer to not reach an error
rs_log_regression = LogisticRegression(max_iter=1500)


from sklearn.model_selection import cross_val_score

k_folds = 10

k3_knn_acc_list = cross_val_score(k3_knn_classifier, HDTFM, LABEL_V, cv = k_folds, scoring ='accuracy')

rs5_dec_tree_acc_list = cross_val_score(rs5_decision_tree, HDTFM, LABEL_V, cv = k_folds, scoring ='accuracy')

rs_log_reg_acc_list = cross_val_score(rs_log_regression, HDTFM, LABEL_V, cv = k_folds, scoring ='accuracy')


acc_cv_k3_knn = k3_knn_acc_list.mean()

acc_cv_dec_tree = rs5_dec_tree_acc_list.mean()

acc_cv_log_reg = rs_log_reg_acc_list.mean()

print ('knn cross validation score: ' + str ( acc_cv_k3_knn) )

print ('decision tree cross validation score: ' + str(acc_cv_dec_tree) )

print ('logistic regression cross validation score: ' + str(acc_cv_log_reg)  )



knn cross validation score: 0.6343010752688172
decision tree cross validation score: 0.750752688172043
logistic regression cross validation score: 0.810752688172043



# Part H Results

Based on running cell H at this moment, these are the results <br>
KNN Cross-Validation score: 0.6343010752688172 <br>
Decision Tree Cross-Validation score: 0.750752688172043 <br>
Logistic Regression Cross-Validation score: 0.810752688172043

